### ISSUE TITLE
Undo behavior is nonsensical

### ISSUE THREAD

**Issue by PedroAlucinante** [jupyter/help#105]
Jupyter appears to implement undo of textual edits on a cell-by-cell basis.   This is extremely confusing, and makes it possible to "undo" to a state the notebook was never actually in, which violates the user's expectation of the undo.  To reproduce:

1. Create a new Python notebook with two cells.
2. Add "1" to the first cell, and add "2" to the second.   Notebook looks like this:

cell 1: 1
cell 2: 2

3. Add the number 3 to the first cell and the number 4 to the second.  Notebook looks like this:

cell 1: 13
cell 2: 24

4. Click in the first cell, and undo twice via the keyboard shortcut.  Notebook now looks like this:

cell 1: [empty]
cell 2: 24

**Comment by takluyver** [comment]
Yes, the undo stack is separate for each cell. Each cell is a separate CodeMirror instance, and it's CodeMirror that provides undo. I'm sure it's a bit odd if you're not used to it, but 'nonsensical' seems a bit harsh.

This might be different in Jupyterlab, I don't know.

**Comment by PedroAlucinante** [comment]
That explains why it is happening, but the behavior is still destructive.

The case on which I encountered it was larger than my example - I had a notebook with ~100 cells, the majority of which contain functions on a one-function-per-cell basis.   I made changes to several of the functions that I wanted to revert, and attempted to do so via undo.   Instead, my code ended up in a state that it had never been in, and I ended up starting all over and losing ~12 hours of work.

"A bit odd if you're not used to it" is a bit of an understatement.  "Undo" is a basic feature present in hundreds of software packages I have used over the course of my life, and there's an expectation of how it works.  If Jupyter is going to deviate from that, it should be obvious that it's happening before it destroys the user's work.

**Comment by PedroAlucinante** [comment]
Is this something that is being looked into?
